### ISSUE TITLE
Event listeners during drag.

### ISSUE THREAD

**Issue by dckesler** [gaearon/react-dnd#321]
It would seem that during drag all other event listeners are stopped. This is probably something you intended, however, I need some way around it.

In our product we are using react-dnd (it's been great) and at one point items that are draggable are also dropTargets. The more obvious issue with this is not wanting an element to be dropped into itself and not wanting the higher dropTarget to handle the drop a second time. Thankfully, both of those problems were easily solved with didDrop from the monitor and comparing id's. The problem is that once I've solved these two issues I'm left with this sort of void where the element I'm dragging is still technically at. If I move an element just slightly it will reset to its original position on account of the first drop being denied because it can't drop in itself and then the higher dropTarget being denied because it thinks that the lower dropTarget handled it already.

On top of that, I felt like maybe I could return "null" if isDragging was true to prevent an element dropping inside itself. Unfortunately, this seems to end the drag event when I'm using a custom drag layer. 

I hope that is a somewhat decent explanation of my issue. So I figured I could just throw in my own drag helper along with a separate way of checking to see if the drop had been accepted the first time. Here's where I ran into real issue. When I add my own event listeners to the window/document they don't fire during the drag with react-dnd. To be more precise, the "mouseup" event never fires, and the "mousemove" event fires like a couple times before stopping.

Just to make sure it wasn't something with my code I went to an example in your docs and threw this into the console:
```javascript
window.addEventListener("mousemove", function(){
  console.log("move")
})
```
It seems to me that the same thing is happening. I see "move" until I start a react-dnd drag, and when I stop it kicks back in. This is exactly the behavior I'd expect from an event stopPropagation() call, but I can't find that anywhere in react-dnd's code. So I've been stuck on this for a good amount of time, and could use some help. I would like to say though that I'm certainly not feeling entitled to any help since the time it would take to rewrite this entire library would be a lot longer than how long I've spent figuring it out.

TL;DR I need to my own event listeners to keep firing during drag.

I did my best searching for a duplicate of this issue, but I obviously could have missed something. Thanks for taking the time in the first place to make and maintain this project, and also if you get around to giving me a nudge in the right direction that would be greatly appreciated.

**Comment by gaearon** [comment]
That `mouse` events don't fire is not an issue with React DnD.

It's just how [HTML5 Drag and Drop API](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Drag_and_drop) works under the hood.
When you use [React DnD HTML backend](https://github.com/gaearon/react-dnd-html5-backend), that's exactly what you're using.

>The problem is that once I've solved these two issues I'm left with this sort of void where the element I'm dragging is still technically at. If I move an element just slightly it will reset to its original position on account of the first drop being denied because it can't drop in itself and then the higher dropTarget being denied because it thinks that the lower dropTarget handled it already.

I don't understand this. :-( Maybe a gif would help?

**Comment by dckesler** [comment]
This is my attempt at a helpful gif.

![dragproblem](https://cloud.githubusercontent.com/assets/8698308/10926616/192eb0ac-8256-11e5-9ce5-0a45eec3bb8f.gif)

So when I drop away from where the square was positioned it's fine. When my mouse is over the area that the square was at then two things happen.
- One, it tries to run the drop function on itself, essentially dropping itself into itself. I stop this by checking to see if the drag item's id is the same as the drop target's id.
- Two, it tries to run the the drop function on the larger element behind it since the square is positioned inside that larger element. This doesn't run because the monitor.didDrop() function says that I've already dropped the square. I need this in other situations because you can actually drop other squares into squares. Like this...

![nesteddrop](https://cloud.githubusercontent.com/assets/8698308/10926695/0415797a-8257-11e5-897f-6901bdae3e77.gif)

So, this is where my idea for a workaround happened. I figured I could have my own `monitor.didDrop()` sort of thing running to check to see if the drop had been actually handled or not. That worked fine, but at that point the second `drop` function can no longer use `monitor.getDifferenceFromInitialOffset()`. So I can't get the difference in mouse positions from the react-dnd stuff. That's when I decided to add my own event listener paired with my own `didDrop` stuff. This way I could track the difference in mouse positions. Here is where I ran into the problem of having my mouse events blocked.

**Comment by gaearon** [comment]
Just to make sure, are you aware that you can return an arbitrary object from `drop()` and it will be available as `monitor.getDropResult()`? This gives you more information than just `didDrop`.

**Comment by dckesler** [comment]
Actually that's just what I needed. I can return the `getDifferenceFromInitialOffset()` in the first drop event and use it in the next one if that one returns null. Thank you for your help.

**Comment by gaearon** [comment]
Great. Glad to have helped.
